Sanowret Crystal script

NBSL opened the pull request:
Script that can be ran in the background to gaze or exhale your crystal. Checks skill state, area restrictions, hiding and concentration levels.

NBSL commented:
This should be my last PR for today. Considering more people have crystals, it might be nice to have a script in the general package.

rcuhljr commented on sanowret-crystal.lic:
```diff
@@ -0,0 +1,39 @@
+no_pause_all
+
+custom_require.call(%w[common drinfomon])
+invalid_room = nil
```
This line doesn't have any effect, you can snip it out.

rcuhljr commented on sanowret-crystal.lic:
```diff
@@ -0,0 +1,39 @@
+custom_require.call(%w[common drinfomon])
+invalid_room = nil
+
+while true do
```
this can just be loop do

rcuhljr commented on sanowret-crystal.lic:
```diff
@@ -0,0 +1,39 @@
+
+while true do
+  line = script.gets?
+  response = ""
```
We could add a line `next if invalid_room == Room.current.id` right here and simplify our later if statement.

rpherbig commented:
I was just about to push my own sanowret crystal script, so it's interesting to compare the two.

I'm concerned that some scripts may not handle well being paused for a long period of time (30-40s when you gaze I think). Have you been running this in the background while in combat or noncombat training?

Is this something that is advantageous to be running in the background? My initial thought was to make a standalone script (for those that want to use before/after) and then also wire it into crossing-training.

rcuhljr commented on sanowret-crystal.lic:
```diff
@@ -0,0 +1,39 @@
+      response = DRC.bput("exhale sanowret crystal", "you come away from the experience with a further understanding of Arcana as the scintillating lights fade again.", "However, nothing much else happens, as you lack the concentration to focus.", "This is not a good place for that.", "Doing that would give away your hiding place.")
+    end
+
+    case response
```
We could use an else statement to handle the first when here without really losing anything other than the giant match statement.
```
when "This is not...
  fail stuff
else
  success stuff
```

rcuhljr commented on sanowret-crystal.lic:
```diff
@@ -0,0 +1,39 @@
+
+    case response
+    when /(?:A soft light blossoms in the very center of the crystal, and begins to fill your mind with the knowledge of \w+.|you come away from the experience with a further understanding of Arcana as the scintillating lights fade again.)/
+      invalid_room = nil
```
There's no reason to clear out the invalid_room id, a bad room number is still bad.

rpherbig commented on sanowret-crystal.lic:
```diff
@@ -0,0 +1,39 @@
+=begin
+	Runs in the background, gazing or exhaling your sanowret crystal. Checks for concentration and current mindstate.
```
We've generally been putting a link to elanthipedia in the header block and documenting things there.

rpherbig commented on sanowret-crystal.lic:
```diff
@@ -0,0 +1,39 @@
+  response = ""
+  if DRStats.concentration == 100 && !hidden? && invalid_room != Room.current.id
+    if DRSkill.getxp('Arcana') <= 10
+      reponse = DRC.bput("gaze sanowret crystal", "A soft light blossoms in the very center of the crystal, and begins to fill your mind with the knowledge of \w+.", "However, nothing much else happens, as you lack the concentration to focus.", "This is not a good place for that.")
```
What if your crystal is not worn?

rcuhljr replied:
There's not likely to be any support here for non worn crystals, the whole point of wearing them is they work during RT and basically no matter what you're doing, not wearing it ruins that.

rcuhljr commented on sanowret-crystal.lic:
```diff
@@ -0,0 +1,39 @@
+      pause 1
+    end
+  end
+  pause 6 unless line
```
Why the six second pause here? I assume we're not waiting on a pulse.

rcuhljr commented:
Is there any concern about it gazing and then six seconds later gazing again or breathing?

rpherbig commented on sanowret-crystal.lic:
```diff
@@ -0,0 +1,39 @@
+=end
+
+no_kill_all
+no_pause_all
```
I _think_ this means that if you are arrested, `jail-buddy` won't be able to pause this script. I don't see any error handling below for the case in which you've temporarily lost your crystal.

NBSL commented:
I've been running it for a few days now, making changes when I run into issues. Only issues I've ran in to was it spamming when I was in dokts, and not being able to exhale while hidden. This has been both in combat and non combat, on both a thief and a trader. So far it handles everything just fine.

As for the not worn aspect, I actually thought about that after I pushed it to the repos. Planned on working on that today and testing it on my barb.

The 6 second wait is based off of an old memory where a tick in DR was 6 seconds. That also is there because I ran into performance issues using 0.5 and 1 second pauses.

The header stuff is because it's on ;repos. I, in truth, didn't think it'd be merged into the default repo but I hadn't seen anyone else post anything about it since that issue that was opened the other day.

And dear lord, you two respond fast.

rcuhljr commented:
Lunch timeeeeee. What kind of performance issues were seeing? we've got plenty of scripts running .1 or lower loops and this guy is doing almost no work.  It might be simpler to just get rid of line.gets? since it's never really used and just always pause one second each loop. If there was an option for this to run once and used stored crystals it would allow us to use the same script for crossing-training and people who like  before/after scripts for training manager.

NBSL commented:
It was freezing the display. Originally I had loop do and was using fputs since the crystals didn't care about most RTs. Earlier this morning I added the hidden check and changed to using bputs. I can move it to a single use if you want.

rpherbig commented on sanowret-crystal.lic:
```diff
@@ -0,0 +1,73 @@
+    @invalid_room = nil
+    args = parse_args(arg_definitions)
+    check_crystal if args.run
+    passive if args.run.nil?
```
You could also `passive unless args.run`

Though I wonder if this could be combined: always go into the loop, exit after the first iteration if the arg was used

NBSL replied:
Is there any way to check if args exist at all? I have a cleanup script I wrote so that I could train alchemy and not end up with tons of trash and misplaced items, but couldn't figure out how to see if the OpenStruct was empty of args. Thats basically where that if was originally from.

rpherbig commented on sanowret-crystal.lic:
```diff
@@ -0,0 +1,73 @@
+
+    case response
+    when "Doing that would give away your hiding place."
+      fput("unhide")
```
We `unhide` in a lot of places. Would you be up for adding a helper to `DRC` that does this (with a bput)? See https://github.com/rpherbig/dr-scripts/blob/1c8306610f09bbfa783d980f41b582c1f21f2b17/crossing-training.lic#L719 for an example

rcuhljr replied:
I'm not keen on unhiding in combat just use the crystal.

NBSL replied:
Sure.

NBSL replied:
@rcuhljr Would you be ok with something like this?
[Unhide?](https://pastebin.com/j0yUbzAt)

Not sure if I need the /.*/ at the end, currently testing it.

rpherbig replied:
Why `assess`?

rcuhljr replied:
No I'd just rather not unhide at all honestly. People hide while doing climbs, if we break that for this and then a mob wanders in and aggros them, yuck. Having to wait while hidden isn't a serious problem, you're not going to clear fully while hidden and if you do so what. Or we get someone killed/afk'd because they hid to go afk and this bumps them out.

NBSL replied:
Outside of introducing events as a require for common, the only other thing I see is is DRRoom.npcs which would return both hostile and non-hostile NPCs. I didn't see anything in lich that would give me current combat status.

NBSL replied:
The other solution could be Script.running?("combat-trainer") But that would stop the passive from running when you're in a crap spawn room like my trader was.

jandersson replied:
Would it make more sense to look at hiding status and not gaze at the crystal if hidden?

NBSL replied:
That's what it was originally @jandersson . This is more in response to adding an unhide check to common.lic

rpherbig commented on sanowret-crystal.lic:
```diff
@@ -0,0 +1,73 @@
+      fput("unhide")
+      use_crystal
+    when "This is not a good place for that."
+      @invalid_room = Room.current.id
```
Why store only the last invalid room? Could this be a list instead?

NBSL replied:
If it's in passive mode, I could have it push to a list. The only time I ran into that error so far, was in with Dokt.

rpherbig commented on sanowret-crystal.lic:
```diff
@@ -0,0 +1,73 @@
+  end
+
+  def passive
+    invalid_room = nil
```
Should this be `@invalid_room`? Does it need to be initialized at all?

NBSL replied:
Ah, the issues editing in 2 places lol. I have it removed in my test one, but forgot to remove it in my forked folder.

rpherbig commented on sanowret-crystal.lic:
```diff
@@ -0,0 +1,73 @@
+  def use_crystal
+    return if @invalid_room == Room.current.id
+    if DRSkill.getxp('Arcana') <= 10
+      reponse = DRC.bput("gaze sanowret crystal", "A soft light blossoms in the very center of the crystal, and begins to fill your mind with the knowledge of \w+.", "However, nothing much else happens, as you lack the concentration to focus.", "This is not a good place for that.")
```
Will actions from other scripts interrupt the gaze? Do we need a pause here?

NBSL replied:
The only issue I've seen is typeahead. I don't know of a way to handle that. 

Other than that, gaze works during RT. Exhale forces you to wait.

NBSL commented:
Ok. I added hidden checks to the spots where I think it'd run into race conditions in passive mode. I changed invalid_room to an array so that while running in passive mode, it'll automatically skip learned rooms.

rpherbig commented on sanowret-crystal.lic:
```diff
@@ -0,0 +1,72 @@
+    ]
+    @invalid_rooms = []
+    args = parse_args(arg_definitions)
+    check_crystal if args.run && !hiding?
```
You don't need to check `hiding?` here, `check_crystal` already does it

rpherbig commented on sanowret-crystal.lic:
```diff
@@ -0,0 +1,72 @@
+
+  def check_crystal
+    return if hiding?
+    if DRStats.concentration == 100 && DRSkill.getxp('Arcana') <= 25
```
I think this function will be a bit clearer if you make this a guard clause (`return if DRStats.concentration != 100 || DRSkill.getxp('Arcana') > 25`)

rpherbig commented on sanowret-crystal.lic:
```diff
@@ -0,0 +1,72 @@
+        fput("get my sanowret crystal")
+        use_crystal
+        fput("stow my sanowret crystal")
+      when /(^You tap a sanowret crystal that you are wearing.$|)/
```
You probably want to escape the `.` here.

Why is this captured (i.e. use of `()`)?

How does the or (`|`) affect this? Won't that only match an empty result?

NBSL replied:
I removed that. It was left over from checking for 2 errors. As of right now I haven't pushed my latest changes, since I ran into issues with everything hanging when I'd try to ambush stun.

rpherbig commented on sanowret-crystal.lic:
```diff
@@ -0,0 +1,72 @@
+      response = DRC.bput("exhale sanowret crystal", "you come away from the experience with a further understanding of Arcana as the scintillating lights fade again.", "However, nothing much else happens, as you lack the concentration to focus.", "This is not a good place for that.", "Doing that would give away your hiding place.")
+    end
+
+    if response =~ "This is not a good place for that."
```
This could also use the guard clause style (`return if response !~ "This is not a good place for that."`)

rpherbig commented on sanowret-crystal.lic:
```diff
@@ -0,0 +1,66 @@
+  end
+
+  def check_crystal
+    return if DRStats.concentration < 100 || DRSkill.getxp('Arcana') >= 25 || hiding?
```
I prefer each guard clause to have its own `return`:
```
return if DRStats.concentration < 100
return if DRSkill.getxp('Arcana') >= 25
return if hiding?
```

rpherbig replied:
But you don't have to change it, just noting a style thing

rpherbig commented on sanowret-crystal.lic:
```diff
@@ -0,0 +1,66 @@
+    return if DRStats.concentration < 100 || DRSkill.getxp('Arcana') >= 25 || hiding?
+    case DRC.bput('tap my sanowret crystal', /^You tap a sanowret crystal inside your .*.$/,/^You tap a sanowret crystal that you are wearing.$/, /^You tap a sanowret crystal that you are holding.$/, "I could not find what you were referring to.")
+    when /You tap a sanowret crystal inside your/
+      fput("get my sanowret crystal")
```
Can you use `bput` (with matcher messages) instead of these `fput`s?
